lib/assembly_report/assembly_report.hpp

```diff
@@ -0,0 +1,148 @@
+
+            return contig_synonyms;
+        }
+
+        public:
+        bool is_contig_available(std::string & contig) {
+            if (index_map.find(contig) != index_map.end()) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        ContigSynonyms get_contig_synonyms(std::string & contig) {
+            if (index_map.find(contig) != index_map.end()) {
+                return contigs[index_map[contig]];
+            } else {
+                throw std::invalid_argument("Contig : " + contig + " not present in synonyms map");
+            }
+        }
+
+        void parse_file(std::istream &report) {
+            std::vector<char> report_line;
+            size_t const default_line_buffer_size = 64 * 1024;
+            size_t const assembly_report_column_count = 10;
+            report_line.reserve(default_line_buffer_size);
+            std::vector<std::string> assembly_report_errors; // to contain errors while parsing
+            std::map<std::string,std::set<int>> multiple_occurance_of_contig_errors;
+
+            int contig_index = 0, line_num = -1;
+            while (util::readline(report, report_line).size() != 0) {
+                line_num++;
+                if (report_line[0] == '#') {
+                    continue;
+                }
+
+                std::vector<std::string> columns;
+                std::string line{report_line.begin(), report_line.end()};
+                boost::algorithm::trim(line);
+                util::string_split(line, "\t", columns);
+                if(columns.size() != assembly_report_column_count) {
+                    std::string error = "Expected " + std::to_string(assembly_report_column_count)
+                                        + " columns, found " + std::to_string(columns.size()) + "\n";
+                    assembly_report_errors.push_back(error);
+                }
+
+                auto contig_synonyms = extract_synonyms(columns);
+                contigs.push_back(contig_synonyms);
+                for (auto contig : contig_synonyms.synonyms) {
+                    if(index_map.find(contig) != index_map.end() && index_map[contig] != contig_index) {
+                        // if a contig is found in two different lines that would be treated as error
+                        multiple_occurance_of_contig_errors[contig].insert(contig_index);
+                        multiple_occurance_of_contig_errors[contig].insert(index_map[contig]);
+                        continue;
+                    }
+                    index_map[contig] = contig_index;
+                }
+                contig_index++;
+
+            }
+
+            for (auto contig_error: multiple_occurance_of_contig_errors) {
+                std::string error = "Contig " + contig_error.first + ", found in multiple lines : ";
+                for (auto line_num : contig_error.second) {
+                    error += std::to_string(line_num) + " ";
```

Comment:
small tip: usually lines are counted in 1-base, that is: the first line is line number 1, even in programming circles (e.g. vim). If you are going to use those line number in messages to the user (and not only for internal counting, where 0-base would be fine), then please count lines in 1-base.

Reply:
agreed

Reply:
agreed,